interfaces/repo_test.go

```
@@ -848,3 +848,179 @@ func (s *RepositorySuite) TestSecuritySnippetsForSnapFailureWithPermanentSnippet
+name: consumer
+apps:
+    app:
+        plugs: [iface]
+plugs:
+    iface:
+        attr: extra
+`
+const testUnknownIfaceConsumerYaml = `
+name: consumer
+apps:
+    app:
+        plugs: [unknown]
+`
+const testInvalidIfaceConsumerYaml = `
+name: consumer
+apps:
+    app:
+        plugs: [invalid]
+`
+const testDifferentLabelConsumerYaml = `
+name: consumer
+apps:
+    app:
+        plugs: [iface]
+plugs:
+    iface:
+        label: New Label
+`
+const testBareConsumerYaml = `
+name: consumer
+apps:
+    app:
+`
+
+const testProducerYaml = `
+name: producer
+apps:
+    app:
+        slots: [iface]
+`
+const testDifferentIfaceProducerYaml = `
+name: producer
+apps:
+    app:
+        slots: [iface]
+slots:
+    iface:
+        interface: other-iface
+`
+const testDifferentAttrsProducerYaml = `
+name: producer
+apps:
+    app:
+        slots: [iface]
+slots:
+    iface:
+        attr: extra
+`
+const testUnknownIfaceProducerYaml = `
+name: producer
+apps:
+    app:
+        slots: [unknown]
+`
+const testInvalidIfaceProducerYaml = `
+name: producer
+apps:
+    app:
+        slots: [invalid]
+`
+const testDifferentLabelProducerYaml = `
+name: producer
+apps:
+    app:
+        slots: [iface]
+slots:
+    iface:
+        label: New Label
+`
+const testBareProducerYaml = `
+name: producer
+apps:
+    app:
+`
+
+func (s *AddRemoveSuite) addSnap(c *C, yaml string) (*snap.Info, error) {
+	snapInfo, err := snap.InfoFromSnapYaml([]byte(yaml))
+	c.Assert(err, IsNil)
+	return snapInfo, s.repo.AddSnap(snapInfo)
+}
+
+func (s *AddRemoveSuite) TestAddSnapAddsPlugs(c *C) {
+	_, err := s.addSnap(c, testConsumerYaml)
+	c.Assert(err, IsNil)
+	// The plug was added
+	c.Assert(s.repo.Plug("consumer", "iface"), Not(IsNil))
+}
+
+func (s *AddRemoveSuite) TestAddSnapPanicsOnExistingSnaps(c *C) {
+	_, _ = s.addSnap(c, testConsumerYaml)
+	_, err := s.addSnap(c, testConsumerYaml)
+	c.Assert(err, ErrorMatches, "cannot add plug consumer.iface: existing plug in the way")
```

Applied in aa2a087fded80dad2bd5fcdc22456e974bf7a475

This also highlights the error swallowing issue I've mentioned above.